### ISSUE TITLE
Further enhancing `talawa_lint`

### ISSUE THREAD

**Issue by literalEval** [PalisadoesFoundation/talawa#1646]
**Is your feature request related to a problem? Please describe.**
Talawa Lint is working great and contributors are getting hold of the new lint rules, but some things can be changed.

1. Exclude some files like `test_helpers.dart`, `test_helper_mocks.dart` etc, which are causing issues to contributors as no one wants to document such huge files and neither does it make sense. Alternatively, we can make another issue for documenting these file to a certain level.
2. Exclude some standard fields from doc check. Functions like `main`, `test` etc.
3. The `params:` and `returns:` blocks should have at least one blank line between them, for them to be shown correctly when user hovers over the corresponding field in the IDE. Although this is an issue related to the IDE itself, still we can fix this on our side. Having a blank line also improves readability.
4. `talawa_good_doc_comments` seems to have trouble with `implicit return types`. We need to fix that.

### Chores

- [x] Exclude test helper files from lint check
- [x] Exclude standard field from doc check
- [x] Treat setters and getters as normal fields.
- [x] Enforce blank line between `params:` and `returns:` blocks.
- [x] Make `params:` and `returns:` directives bold.
- [x] Either enforce explicit return types, which is a better solution for readability, or enhance `talawa_good_doc_comments` to support implicit ones.
- [x] Enforce a tab spaced `None` in `params:` block (currently the tab isn't enforced) 
       **02/04/2023**
- [x] Write a `quick_fix` template that will insert a doc template automatically. 
       **02/04/2023**
- [x] Add hyperlink to `Talawa Docs` in the error itself.  
       **07/04/2023**

### Bugs/Enhancements reported
- [x] Documenting params in wrong order should give a descriptive message. 
       Reported by @Ayush0Chaudhary , 07/04/23

**Comment by literalEval** [comment]
@palisadoes please confirm that this is an issue so that I can start working on it.

**Comment by literalEval** [comment]
@palisadoes I am excluding whole test folder from our custom lint.

Let me explain:

1. Test files mostly have a `main` function and a custom function that returns a widget, and both of these are self explanatory to everyone. Doesn't need any documentation.
2. But what about adding comments to tests so that others can understand what the test is testing ? Sure, but we are not checking these comments in `talawa_lint` anyway, and neither can it be checked. Because we can't judge which line needs a comment and which doesn't. The volunteers and mentors will need to check this.
3. If we don't do this, each test file will contain comments like these

  ```
  /// This is main function.
  void main() {}
  ```

  ```
  /// This returns MyWidget for testing.
  /// 
  /// params:
  /// None
  /// 
  /// returns:
  /// * `MyWidget`: MyWidget
  MyWidget getMyWidget() {
    return MyWidget();
  }

**Comment by palisadoes** [comment]
OK

**Comment by palisadoes** [comment]
@literalEval Is there any more work that needs to be done on this?

**Comment by literalEval** [comment]
yeah @palisadoes . Was busy writing proposal. Will do this by this weekend. Sorry for the delay.

**Comment by literalEval** [comment]
@palisadoes if a function has no parameters and no return value (void return type), should we make the `params:` and `returns` blocks optional ?

**Comment by palisadoes** [comment]
That would mean that no one would document the code if this were optional.
Are there any other ways to handle these cases?

**Comment by literalEval** [comment]
Contributors will still NEED to write documentation for it. That part is mandatory. The only optional part will be the params and returns blocks. The point I am trying to make is that, say there is a function

```dart
void foo() {
    ....stuff...
}
```

In this case the doc looks like

```dart
/// Function that does stuff
///
/// **params**:
///   None
/// **returns**:
///   None
```

As you can see, the params and returns blocks are taking unnecessary space and adding nothing to the documentation.

**Comment by palisadoes** [comment]
The issue is that because it's optional people will just put a single line for everything.

So the overall quality of the documentation is now suffering because of a minority of functions and methods that require a little extra text. 

This is not  a good development.

**Comment by palisadoes** [comment]
All the work you have put in to fix the documentation via the linter will be circumvented by others when they realize they don't have to do this work.

**Comment by literalEval** [comment]
@palisadoes understood. Moreover, the original tasks of this issue are completed, but more will follow. So instead of closing this issue, let's keep it opened and add new tasks to it as they come. This will save us from opening unnecessary issues of the same kind. If you don't agree with this point, please feel free to close this :)

**Comment by literalEval** [comment]
@palisadoes when was the `docs` directory in the Talawa App codebase last updated ?

**Comment by palisadoes** [comment]
Open an issue to fix this

**Comment by literalEval** [comment]
Sure.

**Comment by literalEval** [comment]
@palisadoes can you think of anything extra that we can add to our lint system? Or should I halt it's further development for now and work on something else? :)

**Comment by Ayush0Chaudhary** [comment]
@literalEval did you fix all the wrong error handling ?

**Comment by literalEval** [comment]
@Ayush0Chaudhary see the PR just above. I have updated the error message. Please advise if it needs more updation,

**Comment by palisadoes** [comment]
What's happening with fixing the original linter package?

**Comment by literalEval** [comment]
@palisadoes I am having discussions with the plugin maintainer on Discord. The main issue at that time was writing tests for the changes. I think we can do that by collaborating with each other. Maybe I can write the changes and @Ayush0Chaudhary can write the tests ?

**Comment by palisadoes** [comment]
That's a possibility, but you'll need to ask @Ayush0Chaudhary about his availability and willingness to do this.

Why hasn't there an update here? The tests he's suggesting don't look time consuming.

- https://github.com/invertase/dart_custom_lint/pull/82

**Comment by Ayush0Chaudhary** [comment]
@palisadoes @literalEval @noman2002 
I am glad you asked and will be happy to work with you @literalEval , but right now I am focusing on bringing back all broken features through  https://github.com/PalisadoesFoundation/talawa/issues/1748 


I would need 4 Days to fix that issue

**Comment by literalEval** [comment]
@palisadoes the original PR was easy enough to write tests for. But then the author advised me to rewrite the implementation and then write tests accordingly. I am discussing the new implementation with him and start working on it soon. 

@Ayush0Chaudhary please let me know when you are free. Thanks.

**Comment by literalEval** [comment]
@palisadoes please keep this opened. Though no bug has been found in our custom linter as of now, it will be good if this issue exists as a parent issue for any further enhancement in our linter. Thanks.

**Comment by palisadoes** [comment]
@literalEval 

Are you still working on this?

**Comment by literalEval** [comment]
> @palisadoes please keep this opened. Though no bug has been found in our custom linter as of now, it will be good if this issue exists as a parent issue for any further enhancement in our linter. Thanks.

This was the reason for keeping this issue open.

**Comment by aikansh008** [comment]
please assign this issue to me

**Comment by dhanagopu** [comment]
@aikansh008  Our policy is to assign no more than two issues to each contributor across all repositories. This way everyone gets a chance to participate in the projects. We sometimes give exceptions for more urgent cases and sometimes we lose track, but the policy stands. You have reached your limit, please wait until your existing issues are closed before requesting more issues. You could unassign yourself from one of the other issues too.

**Comment by palisadoes** [comment]
Unassigning due to inactivity and/or no submitted PR. Others need to get a chance. Please let us know whether you want to be reassigned before someone else requests assignment or select another issue.
